presto-main/src/test/java/io/prestosql/sql/planner/TestUnwrapCastInComparison.java

```diff
@@ -0,0 +1,724 @@
+                                values("A"))));
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = DOUBLE '1.9'",
+                anyTree(
+                        filter("A IS NULL AND NULL",
+                                values("A"))));
+
+        // below top of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = DOUBLE '32766'",
+                anyTree(
+                        filter("A = SMALLINT '32766'",
+                                values("A"))));
+
+        // round to top of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = DOUBLE '32766.9'",
+                anyTree(
+                        filter("A IS NULL AND NULL",
+                                values("A"))));
+
+        // top of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = DOUBLE '32767'",
+                anyTree(
+                        filter("A = SMALLINT '32767'",
+                                values("A"))));
+
+        // above range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = DOUBLE '32768'",
+                anyTree(
+                        filter("A IS NULL AND NULL",
+                                values("A"))));
+
+        // above bottom of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = DOUBLE '-32767'",
+                anyTree(
+                        filter("A = SMALLINT '-32767'",
+                                values("A"))));
+
+        // round to bottom of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = DOUBLE '-32767.9'",
+                anyTree(
+                        filter("A IS NULL AND NULL",
+                                values("A"))));
+
+        // bottom of range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = DOUBLE '-32768'",
+                anyTree(
+                        filter("A = SMALLINT '-32768'",
+                                values("A"))));
+
+        // below range
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a = DOUBLE '-32769'",
+                anyTree(
+                        filter("A IS NULL AND NULL",
+                                values("A"))));
+    }
+
+    @Test
+    public void testNotEquals()
+    {
+        // representable
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a <> DOUBLE '1'",
+                anyTree(
+                        filter("A <> SMALLINT '1'",
+                                values("A"))));
+
+        // non-representable
+        assertPlan(
+                "SELECT * FROM (VALUES SMALLINT '0') t(a) WHERE a <> DOUBLE '1.1'",
+                anyTree(
+                        filter("NOT (A IS NULL) OR NULL",
```

Comment:
Would the test pass if you had `A IS NOT NULL` here?

Reply:
Future TODO: we should have a rule that realizes this is as good as `filter("A IS NOT NULL")`. Removing `OR` would help predicate pushdown.

Reply:
> Would the test pass if you had A IS NOT NULL here?

No, the planner desugars `A IS NOT NULL` to `NOT (A IS NULL)`